src/clockcache.c

```diff
@@ -1841,11 +1843,12 @@ clockcache_init(clockcache          *cc,   // OUT
 
    /* Entry per-thread ref counts */
    size_t refcount_size = cc->cfg->page_capacity * CC_RC_WIDTH * sizeof(uint8);
-   cc->rc_bh = platform_buffer_create(refcount_size, cc->heap_handle, mid);
-   if (!cc->rc_bh) {
+
+   rc = platform_buffer_init(&cc->rc_bh, refcount_size);
+   if (!SUCCESS(rc)) {
```

Comment:
The error handling needs to be tightened up.

If either `platform_buffer_init` fails, then we `goto alloc_error`, which does `clockcache_deinit`.  And `clockcache_deinit` unconditionally calls `platform_buffer_deinit` on both buffers.  And `platform_buffer_deinit` unconditionally calls `munmap` on the `buffer_handle`'s `addr`.  But, depending on where we fail, none, one, or both of the buffers may have not been inited, so calling `munmap` could cause an error.

There are two ways to fix this, I think:

1. Modify `clockcache_deinit` to check whether the corresponding pointer is `NULL` before it calls `platform_buffer_deinit`.  So, it would check `cc->data` before calling `platform_buffer_deinit(&cc->bh)` and it would check `cc->refcount` before calling `platform_buffer_deinit(&cc->rc_bh)`.
2. Modify `clockcache_init` to have different error handling targets and do the cleanup on its own, i.e. have `goto alloc_failure1;`, `goto alloc_failure2;`, etc.

My preference is Option 1, but either is fine.

Reply:
@rtjohnso - There are quite a few callers to `platform_buffer_deinit()`, and technically, one has to comb through all instances to make sure that we are not calling this wrongly.

A 3rd approach would be to push the checks to inside of `platform_buffer_deinit()`.

It will try to do the `munmap()` only if `bh->addr` is non-NULL. This will solve the error handling issue from `clockcache_deinit()` and also protect all other instances where `platform_buffer_deinit()` is being called.

What are your thoughts on this approach (before I implement it).

Reply:
(And, ooops, ... sorry, I introduced this error handling issue when implementing the change to introduce `init() / deinit()` interface, without looking too hard at error handling. Sorry!)

Reply:
@rtjohnso : Re: your comment that "_Before the buffer_handle is inited, addr may be garbage, so you cannot safely deinit based on the value of addr._"

In the two places we are talking about, both `rc_allocator_init()` and `clockcache_init()` do a `ZERO_CONTENTS(al)` and `ZERO_CONTENTS(cc)`.

So, certainly for the error path going through `clockcache_deinit(), we should be sure that `cc->bh->addr` will be NULL even before `init()` happened.

Unless you can spot something else I am overlooking, seems like pushing the NULL-check to inside `platform_buffer_deinit()` should work.

Reply:
Discussed this in PR meeting. Rework this as suggested by @robj. `clockcache_deinit()` will only call `platform_buffer_deinit()` only if corresponding pointers are non-NULL.

Remove the fix as applied in the latter function.